src/OrchardCore.Modules/OrchardCore.Media/Processing/MediaTokenService.cs

```diff
@@ -98,30 +131,97 @@ public bool TryValidateToken(IDictionary<string, string> commands, string token)
+        {
+            using var builder = ZString.CreateStringBuilder();
+            builder.Append(TokenCacheKeyPrefix);
+            foreach (var pair in values)
+            {
+                builder.Append(pair.Value.ToString());
+            }
+            return builder.ToString();
+        }
+
+        private static string CreateQueryStringTokenKey(IDictionary<string, string> values)
         {
             using var builder = ZString.CreateStringBuilder();
             builder.Append(TokenCacheKeyPrefix);
-            foreach (var item in values)
+            foreach (var pair in values)
             {
-                builder.Append(item);
+                builder.Append(pair.Value);
             }
             return builder.ToString();
         }
 
         private string GetHash(string queryStringTokenKey)
-            => _memoryCache.GetOrCreate(queryStringTokenKey, entry =>
-               {
-                   entry.SlidingExpiration = TimeSpan.FromHours(5);
+        {
+            if (!_memoryCache.TryGetValue(queryStringTokenKey, out var result))
+            {
+                using var entry = _memoryCache.CreateEntry(queryStringTokenKey);
+
+                entry.SlidingExpiration = TimeSpan.FromHours(5);
+
+                using var hmac = new HMACSHA256(_hashKey);
+
+                // queryStringTokenKey also contains prefix
+                var chars = queryStringTokenKey.AsSpan(TokenCacheKeyPrefix.Length);
 
-                   using var hmac = new HMACSHA256(_hashKey);
+                // only allocate on stack if it's small enough
+                var requiredLength = Encoding.UTF8.GetByteCount(chars);
+                var stringBytes = requiredLength < 1024
+                    ? stackalloc byte[requiredLength]
+                    : new byte[requiredLength];
 
-                   // queryStringTokenKey also contains prefix
-                   var bytes = Encoding.UTF8.GetBytes(queryStringTokenKey, TokenCacheKeyPrefix.Length, queryStringTokenKey.Length - TokenCacheKeyPrefix.Length);
+                // 256 for SHA-256, fits in stack nicely
+                Span<byte> hashBytes = stackalloc byte[hmac.HashSize];
 
-                   var hash = hmac.ComputeHash(bytes);
+                var stringBytesLength = Encoding.UTF8.GetBytes(chars, stringBytes);
 
-                   return Convert.ToBase64String(hash);
-               });
+                hmac.TryComputeHash(stringBytes.Slice(0, stringBytesLength), hashBytes, out var hashBytesLength);
+
+                entry.Value = result = Convert.ToBase64String(hashBytes.Slice(0, hashBytesLength));
+            }
+
+            return (string) result;
+        }
+
+        /// <summary>
+        /// Custom version of <see cref="QueryHelpers.AddQueryString(string,string,string)"/> that takes our pre-built
+        /// dictionary, uri as ReadOnlySpan&lt;char&gt; and uses ZString. Otherwise same logic.
+        /// </summary>
+        private static string AddQueryString(
```

Comment:
this ones copes with `ReadOnlySpan<char>` and uses `ZString` instead of allocating `StringBuilder`